packages/datastore/src/listfield.ts

```diff
@@ -525,23 +539,50 @@ namespace Private {
 
     // Iterate over the identifiers to insert.
     while (i < n) {
-      // Find the boundary identifier for the current id.
-      let j = ArrayExt.lowerBound(metadataIds, ids[i], StringExt.cmp);
-
-      // Bail early if the boundary is the end of the array.
-      if (j === metadataIds.length) {
-        chunks.push({ index: j, ids: ids.slice(i), values: values.slice(i) });
-        break;
+      // If the id has been concurrently deleted, update the cemetery
+      // and continue processing ids.
+      if (checkCemeteryForInsert(ids[i], metadata.cemetery)) {
+        i++;
+        continue;
       }
 
+      // Find the boundary identifier for the current id.
+      let j = ArrayExt.lowerBound(metadata.ids, ids[i], StringExt.cmp);
+
       // Set up the chunk ids.
       let chunkIds: string[] = [];
 
       // Set up the chunk values.
       let chunkValues: T[] = [];
 
-      // Find the extent of the chunk.
-      while (i < n && StringExt.cmp(ids[i], metadataIds[j]) < 0) {
+      // Bail early if the boundary is the end of the array.
```

Comment:
Can you explain a bit what is going on here? I'm wondering if we can simplify this in any way.

Reply:
Essentially, we have to check the cemetery for *every* inserted metadata ID to see if it should be added or not. So if we are inserting a chunk, it's not enough to just check the leading ID.

So the changed logic here is to check all every the cemetery for every ID. The actual inserted chunks that result from this function should not be different, except that they will be filtered of IDs that are in the cemetery.